Fixed #36070 -- Doc'd model validation behavior for composite pks. (#19022)
django/django: docs/topics/composite-primary-key.txt

however, defining a primary key consisting of multiple fields is sometimes
necessary.

> To use a composite primary key, when creating a model set the ``pk`` property

Reviewer: ```suggestion
To use a composite primary key, when defining a model set the ``pk`` attribute
```

    >>> item.pk
    (1, "A755H")

> You can assign a ``tuple`` to the :attr:`~django.db.models.Model.pk` property.

Reviewer 2: Unsure if we should refer to this as a `pk` field or a property, as although I get that you're referencing is as a property here due to the docs link, elsewhere we refer to it as a field

Author: I wondered about this also, but throughout this [section](https://docs.djangoproject.com/en/4.2/ref/models/instances/#the-pk-property) it says "property". WDYT?

Author: > elsewhere we refer to it as a field

I think the main one in conflict is the instruction to set the CompositePrimaryKey to pk:
>  when creating a model set the ``pk`` field 

I would suggest changing that to property. Then it's more consistent, and helps you grok the rest of the page which explains all the ways `pk` does _not_ act like a field for form and model validation and relations.

(Then there is a reference to the `models.CompositePrimaryKey(Field)` as a virtual field, which it is, but that's not inconsistent with `pk` being a property.)

Reviewer 2: Sounds good, we should also update the release note as it also says "set the pk field"

Reviewer 2: There's one more field reference in the `CompositePrimaryKey` definition we might want to change:
```
This field must be defined as the model's ``pk`` field.
```

I think I want another pair of eyes in case other folks have good ideas (cc @csirmazbendeguz @LilyFoote @charettes )

Reviewer: Maybe we should use `attribute` here to disambiguate?

Setting the primary composite field ``pk`` as a form field raises an unknown
field :exc:`.FieldError`.

Composite primary keys in model validation
==========================================

> Since ``pk`` is only a virtual field, the ``exclude`` arguments of
> :ref:`model validation <validating-objects>` methods
> :meth:`.Model.full_clean` and :meth:`.Model.clean_fields` do not recognize
> them. To exclude the composite primary key fields from model validation,
> specify each field individually.

Reviewer 2: ```suggestion
Since ``pk`` is only a virtual field, including ``pk`` as a field name in the
``exclude`` argument :meth:`.Model.clean_fields` has no effect. To exclude the
composite primary key fields from :ref:`model validation <validating-objects>`,
specify each field individually.
```
Maybe?
As `full_clean` calls `validate_unique`, then I guess passing pk in still has an impact

Author: Ah of course!

:meth:`.Model.validate_unique` can still be called with ``exclude={"pk"}`` to
skip uniqueness checks.

> .. admonition:: Primary key fields are read only

Reviewer 2: I think this admonition needs to be in the section above

    This is also true of composite primary keys. Hence, you may want to set
    :attr:`.Field.editable` to ``False`` on all primary key fields to exclude
    them from ModelForms.

Composite primary keys in model validation
==========================================

Since ``pk`` is only a virtual field, including ``pk`` as a field name in the

> ``exclude`` argument of :meth:`.Model.clean_fields` has no effect. To exclude

Reviewer 3: Is this consistent behaviour with other virtual fields? Would it be difficult to support virtual fields in `exclude`?

Author: Supporting was something I mentioned when raising the ticket, but Sarah helpfully pointed out that generated fields are also excluded from `exclude`. M2M fields are another story: they don't appear in `self._meta.fields`, so `clean_fields` never fathoms them anyway & wouldn't need to be excluded.
